# Computer Graphik: Projektive Geometrie I - Übung

I.BA_CG, SW 03

Léonard Wyrsch

In [1]:
import math

## 1. Rechnen mit Vektoren

### (a) Wie lautet der Einheitsvektor n, der die zum Vektor a = \[1,−4,3\]<sup>T</sup> entgegengesetzte Richtung hat?

         1
    a = -4              n = norm(-a)
         3

        -1
    -a = 4              n = -a / |-a|
        -3

    |-a| = |a| = √(a^T * a) = √(1^2 + (-4)^2 + 3^2) = √(1 + 16 + 9) = √(26)

       -1 / √(26)
    n = 4 / √(26)
       -3 / √(26)

In [2]:
print([x / math.sqrt(26) for x in [-1, 4, -3]])

[-0.19611613513818404, 0.7844645405527362, -0.5883484054145521]


       -0.19611613513818404
    n = 0.7844645405527362
       -0.5883484054145521

### (b) Welchen Winkel schliessen die Vektoren a = \[10,−5,10\]<sup>T</sup> und b = \[6,−2,−1\]<sup>T</sup> ein?

    a • b = |a| * |b| * cos(φ)

    a • b = 10 * 6 + (-5) * (-2) + 10 * (-1) = 60 + 10 - 10 = 60
    |a| = √(a • a) = √(100 + 25 + 100) = √(225)
    |b| = √(36 + 4 + 1) = √(41)

    60 / (√(225) * √(41)) = cos(φ)
    φ = cos^(-1)(60 / (√(225) * √(41)))

In [3]:
angle_rad = math.acos(60 / (math.sqrt(225) * math.sqrt(41)))
angle_deg = angle_rad / (2 * math.pi) * 360

print(angle_rad)
print(angle_deg)

0.8960553845713439
51.340191745909905


    φ = 51.340191745909905°

### (c) Zeigen Sie, dass die beiden Vektoren a = \[−1,2,5\]<sup>T</sup> und b = \[−4,8,−4\]<sup>T</sup> zueinander orthogonal sind

    orthogonal => a • b = 0

    a • b = (-1) * (-4) + 2 * 8 + 5 * (-4) = 4 + 16 + (-20) = 20 - 20 = 0    Q.E.D

### (d) Berechnen Sie die Komponente von des Vektors b = \[5,1,3\]<sup>T</sup> in Richtung des Vektors a = \[2,−2,1\]<sup>T</sup>.

b<sub>a</sub> = a • b / |a|^2 * a

    a • b = 10 - 2 + 3 = 11
    |a| = √(4 + 4 + 1) = √9

    a • b / |a|^2 = 11 / 9

In [4]:
print([x * 11 / 9 for x in [2, -2, 1]])

[2.4444444444444446, -2.4444444444444446, 1.2222222222222223]


b<sub>a</sub> =

     2.4444444444444446
    -2.4444444444444446
     1.2222222222222223

### (e) Gegeben sind die Vektoren a = \[1,4,−6\]<sup>T</sup>, b = \[2,−1,2\]<sup>T</sup> und c = \[0,2,3\]<sup>T</sup>. Berechnen Sie (a − b) × (3c).

            1    2  -1
    a - b = 4 - -1 = 5
           -6    2  -8

         3 * 0   0
    3c = 3 * 2 = 6
         3 * 3   9

    -1   0    5 * 9 - -8 * 9   45 + 72  117
     5 x 6 = -8 * 0 - -1 * 9 =  0 +  9 =  9
    -8   9   -1 * 6 -  5 * 0   -6 -  0   -6

                    117
    (a - b) x (3c) =  9
                     -6

### (f) Wie muss der Parameter λ gewählt werden, damit die drei Vektoren a = \[1,λ,4\]<sup>T</sup>, b = \[−2,4,11\]<sup>T</sup> und c = \[−3,5,1\]<sup>T</sup> komplanar sind?

                 1 λ  4
    [a, b, c] = -2 4 11 = 0
                -3 5  1

    [a, b, c] = 1*4*1 + λ*11*-3 + 4*-2*5 - 4*4*-3 - λ*-2*1 - 1*11*5 =
    = 4 - 33λ - 40 + 48 + 2λ - 55 = -31λ - 43 = 0       | +31λ
    -> 31λ = -43                                        | /31
    -> λ = -43/31

In [5]:
print(-43 / 31)

-1.3870967741935485


    λ = -1.3870967741935485

### (g) Zeigen Sie die lineare Unabhängigkeit der Vektoren a = \[1,−6,−4\]<sup>T</sup>, b = \[1,−2,−2\]<sup>T</sup> und c = \[1,2,3\]<sup>T</sup>.

                1 -6 -4
    [a, b, c] = 1 -2 -2 != 0
                1  2  3

    [a, b, c] = 1*-2*3 + -6*-2*1 + -4*1*2 - -4*-2*1 - -6*1*3 - 1*-2*2 =
    = -6 + 12 - 8 - 8 + 18 + 4 = 12 != 0    Q.E.D

### (h) Beweisen Sie den Kosinussatz c<sup>2</sup> = a<sup>2</sup> + b<sup>2</sup> − 2ab*cos(γ) indem Sie im Dreieck mit den Ecken A, B, C die drei Vektoren a = CA, b = CB und c = AB einführen und dann das Skalarprodukt c • c berechnen.
Tipp: es gilt c = a − b und weiter für einen beliebigen Vektor x, x • x = |x|<sup>2</sup> = x<sup>2</sup> wobei x = |x| die Länge des Vektors x ist.

    c • c = |c|^2 = c^2
    c = a - b
    |c|^2 = |a|^2 + |b|^2 - 2*|a|*|b|*cos(γ)

## 2. Skalar- und Vektorprodukt
### Gegeben sei eine Gerade x(t) = x<sub>0</sub> + t*d, t ∈ R und eine Ebene gegeben durch einen Punkt y<sub>0</sub> und den Normalenvektor n. Was sind die Bedingungen dafür, dass
### (a) die Gerade senkrecht auf der Ebene steht,

    d x n = 0

### (b) die Gerade parallel zur Ebene verläuft,

    d • n = 0

### (c) die Gerade in der Ebene liegt?

    d • n = 0 && n • (x0 - y0) = 0

## 3. Translation im 2D
### Das Dreieck A(0,0), B(4,0), C(2,3) werde um den Vektor (1,2)<sup>T</sup> translatiert. Was sind die neuen Koordinaten der drei Eckpunkte? Ist es möglich diese Transformation mit Hilfe einer 2 × 2-Matrix zu beschreiben? Zeichnen Sie die Situation auf!

        0       4       2       1 0 1
    A = 0   B = 0   C = 3   T = 0 1 2
        1       1       1       0 0 1
        

                0       4       2
            A = 0   B = 0   C = 3
                1       1       1

        1 0 1   1       5       3
    T = 0 1 2   2       2       5
        0 0 1   1       1       1

    A' = (1, 2)
    B' = (5, 2)
    C' = (3, 5)

## 4. Translation im 3D
### Das Dreieck A(0,0,0), B(4,0,2), C(2,3,1) werde um den Vektor (1,2,3)<sup>T</sup> translatiert. Was sind die neuen Koordinaten der drei Eckpunkte?

    A' = (1, 2, 3)
    B' = (5, 2, 5)
    C' = (3, 5, 4)

## 5. Skalierung im 2D
### Das Dreieck A(0,0), B(4,0), C(2,3) werde in x-Richtung mit dem Faktor s<sub>x</sub> = 2 und in y-Richtung mit dem Faktor s<sub>y</sub> = 1 skaliert. Bestimmen Sie die Skalierungsmatrix und berechnen Sie die skalierten Eckpunkte. Verwenden Sie Matrix-Vektor-Multiplikationen! Skizzieren Sie die Situation zur Kontrolle.

    S = 2 0
        0 1


            A = 0   B = 4   C = 2
                0       0       3

    S = 2 0     0       8       4
        0 1     0       0       3

    A' = (0, 0)
    B' = (4, 0)
    C' = (4, 3)

## 6. Rotation im 2D
### Das Dreieck A(0,0), B(4,0), C(2,3) werde um den Ursprung gedreht und zwar mit dem Winkel φ = 135°. Bestimmen Sie die Rotations- oder Drehmatrix. Bestimmen Sie die Eckpunkte des rotierten Dreiecks mit Hilfe von Matrix-Vektor-Multiplikationen! Skizzieren Sie die Situation zur Kontrolle.

    R = cos(φ) -sin(φ) = -1/√2 -1/√2
        sin(φ)  cos(φ)    1/√2 -1/√2


                    A = 0   B = 4   C = 2
                        0       0       3

    R = -1/√2 -1/√2     0    -4/√2    -5/√2
         1/√2 -1/√2     0     4/√2    -1/√2

    A' = (    0,     0)
    B' = (-4/√2,  4/√2)
    C' = (-5/√2, -1/√2)

## 7. Nochmals Rotation im 2D
### Wie lautet die Drehmatrix, um das in der vorigen Aufgabe rotierte Dreieck wieder zurück zu drehen? Wie hängen die beiden letzten Drehmatrizen zusammen?

    φ = -135°

    R' = -1/√2  1/√2
         -1/√2 -1/√2


    R * R' =              -1/√2         1/√2
                          -1/√2        -1/√2

    -1/√2 -1/√2     2*(-1/√2)^2            0 = 2*(1/2)        0 = 1 0
     1/√2 -1/√2               0  2*(-1/√2)^2         0  2*(1/2)   0 1

    R * R^(-1) = I

## 8. Spiegelung an der x-Achse
### Spiegeln Sie das Dreieck A(0,0), B(4,0), C(2,3) an der x-Achse. Wie muss eine entsprechende Matrix aussehen, die das bewirkt? Wie sieht die Matrix für die Spiegelung an der y-Achse aus? Wie für eine Spiegelung an einer beliebigen Geraden durch den Ursprung?

            A = 0   B = 4   C = 2
                0       0       3
    
    Mx = 1  0   0       4       2
         0 -1   0       0      -3

    Ax = (0,  0)
    Bx = (4,  0)
    Cx = (2, -3)


            A = 0   B = 4   C = 2
                0       0       3
    
    My = -1 0   0      -4      -2
          0 1   0       0       3

    Ax = ( 0, 0)
    Bx = (-4, 0)
    Cx = (-2, 3)